File: cpp/002.cpp
/**
 *  * Definition for singly-linked list.
 *   * struct ListNode {
 *    *     int val;
 *     *     ListNode *next;
 *      *     ListNode(int x) : val(x), next(NULL) {}
 *       * };
 *        */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *dummy = new ListNode(0), *tail = dummy;
        int carry = 0;
        while (l1 && l2) {
            int temp = carry + l1->val + l2->val;
            tail->next = new ListNode(temp%10);
            carry = temp / 10;
            l1 = l1->next;
            l2 = l2->next;
            tail = tail->next;
        }
        while (l1) {
            int temp = carry + l1->val;
            tail->next = new ListNode(temp%10);
            carry = temp / 10;
            l1 = l1->next;
            tail = tail->next;
        }
        while (l2) {
            int temp = carry + l2->val;
            tail->next = new ListNode(temp%10);
            carry = temp / 10;
            l2 = l2->next;
            tail = tail->next;
        }
        if (carry)
            tail->next = new ListNode(carry);
        return dummy->next;
    }
};

File: go/002.go
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    dummy, carry := &ListNode{}, 0
    tail := dummy
    for l1 != nil && l2 != nil {
        temp := l1.Val + l2.Val + carry
        tail.Next = &ListNode{temp % 10, nil}
        carry = temp / 10
        l1, l2, carry, tail = l1.Next, l2.Next, temp/10, tail.Next
    }
    for l1 != nil {
        temp := l1.Val + carry
        tail.Next = &ListNode{temp % 10, nil}
        carry = temp / 10
        l1, carry, tail = l1.Next, temp/10, tail.Next
    }
    for l2 != nil {
        temp := l2.Val + carry
        tail.Next = &ListNode{temp % 10, nil}
        carry = temp / 10
        l2, carry, tail = l2.Next, temp/10, tail.Next
    }
    if carry > 0 {
        tail.Next = &ListNode{carry, nil}
    }
    return dummy.Next
}
